pkg/adapter/adapter.go

```diff
@@ -31,17 +35,150 @@ type envConfig struct {
+	namespace  string
+	ce         cloudevents.Client
+	client     *mongo.Client
+	source     string
+	datasource *dataSource
+	sinkurl    string
+	logger     *zap.SugaredLogger
+}
+
+// dataSource interface to interact with either a database or a collection.
+type dataSource interface {
+	Watch(ctx context.Context, pipeline interface{}, opts ...*options.ChangeStreamOptions) (*mongo.ChangeStream, error)
+}
+
+// NewEnvConfig creates an empty environement variables configuration.
 func NewEnvConfig() adapter.EnvConfigAccessor {
 	return &envConfig{}
 }
 
 // NewAdapter creates an adapter to convert incoming MongoDb changes events to CloudEvents and
-// then sends them to the specified Sink
+// then sends them to the specified Sink.
 func NewAdapter(ctx context.Context, processed adapter.EnvConfigAccessor, ceClient cloudevents.Client) adapter.Adapter {
-	fmt.Printf("Hey from mongodb adapter")
-	time.Sleep(600 * time.Second)
+	logger := logging.FromContext(ctx)
+	env := processed.(*envConfig)
+
+	rawURI, err := ioutil.ReadFile(env.MongoDbCredentialsPath + "/URI")
+	if err != nil {
+		logger.Desugar().Error("Unable to get MongoDb URI field", zap.Any("secretPath", env.MongoDbCredentialsPath+"/URI"))
+	}
+	URI := string(rawURI)
+	if URI == "" {
+		logger.Desugar().Error("MongoDb URI field is an empty string", zap.Any("URI", URI))
+	}
+
+	return newAdapter(ctx, env, ceClient, URI)
+}
+
+func newAdapter(ctx context.Context, env *envConfig, ceClient cloudevents.Client, URI string) adapter.Adapter {
+	logger := logging.FromContext(ctx)
+
+	// Create new Client.
+	client, err := mongo.NewClient(options.Client().ApplyURI(URI))
+	if err != nil {
+		logger.Desugar().Error("Error creating mongo client", zap.Error(err))
+	}
+	// Get dataSource
+	dataSource := getDataSource(client, env.Database, env.Collection)
+
+	return &mongoDbAdapter{
+		namespace:  env.Namespace,
+		ce:         ceClient,
+		client:     client,
+		datasource: dataSource,
+		source:     env.EventSource,
+		sinkurl:    env.SinkURL,
+		logger:     logger,
+	}
+}
+
+// Returns the appropriate dataSource: database or collection.
+func getDataSource(client *mongo.Client, db string, coll string) *dataSource {
+	if coll != "" {
+		var collsource dataSource = client.Database(db).Collection(coll)
+		return &collsource
+	}
+	var dbsource dataSource = client.Database(db)
+	return &dbsource
+}
+
+// Start creates the watch stream that will watch for dataSource changes.
+func (a *mongoDbAdapter) Start(ctx context.Context) error {
+	// Connect to Client.
+	err := a.client.Connect(ctx)
+	if err != nil {
+		a.logger.Desugar().Error("Error connecting to database", zap.Error(err))
+		return err
+	}
+	defer a.client.Disconnect(ctx)
+
+	// Create a watch stream for either the database or collection.
+	var datasource dataSource = *a.datasource
+	stream, err := datasource.Watch(ctx, mongo.Pipeline{})
+
+	if err != nil {
+		a.logger.Desugar().Error("Error setting up changeStream", zap.Error(err))
```

Comment:
Same. Check if the called already logs.

Reply:
same, returns some errors but doesnt log

Reply:
Same as above